logmerge.man
.TH LOGMERGE 1 "4 July 2006" Tardis "Logmerge manual"
.SH NAME
logmerge \- merge dated log files
.SH SYNOPSIS
.B logmerge
.RB [ \-\-no\-update ]
.RB [ \-\-offsets
.IR file ]
.SH DESCRIPTION
Merge several log files together line by line, keeping the lines in
chronological order.  The position after the last message read from each file
is stored so that messages are not duplicated in subsequent runs.  Information
about the log files is stored in the
.I offsets
file.  A new
.I offsets
file can be produced using the
.B logmerge-mkoffsets
(1) program.
.SH OPTIONS
.IP \-\-no\-update
Do not update the
.I offsets
file with the last position read from each log file.  The next time the program
is run all of the output will be repeated.  This is useful for testing.
.IP "\-\-offsets file"
Use the given file name rather than the default of
.IR /var/lib/logmerge/offsets .
.SH EXAMPLE
The principle intended use for
.B logmerge
is to combine log files for regular summary reports generated by
.IR piperlog (1).
First, make an initial
.I offsets
file:
.IP
logmerge-mkoffsets /var/log/auth.log /var/log/syslog > /var/lib/logmerge/offsets
.PP
then set up a cron job to produce the summaries at regular intervals:
.IP
5 */4 * * * (logmerge | piperlog | mail -s "Log summary at `date`" recipient@example.com)
.SH LIMITATIONS
Log rotation may change the file names, so
.B logmerge
uses each file's inode number (its raw identity in the file system) to find the
correct file to start reading from.  For example, if the log file is called
.I syslog
is rotated to
.I syslog.0
then
.B logmerge
notices that 
.I syslog
now has a different inode number and so checks
.IR syslog.0 .
.PP
Strange file systems may not preserve inode numbers. This will cause the
entire log file to be output each time
.B logmerge
is run.
.PP
The messages in the log files are assumed to be in chronological order.  If
they are not then all the lines will still be output, but not necessarily in
the correct order.  Year rollovers may not yet be handled correctly.
.SH FILES
.IR /var/lib/logmerge/offsets .
.SH AUTHORS
Brian Campbell (bacam@tardis.ed.ac.uk).
.SH SEE ALSO
.BR logmerge-mkoffsets (1),
.BR piperlog (1)


debian/changelog
logmerge (0.1+darcs) unstable; urgency=low

  * Some working version.

 -- Brian Campbell <bacam@tardis.ed.ac.uk>  Fri, 15 Dec 2006 21:57:00 +0000

logmerge (0.1) unstable; urgency=low

  * Initial Release.

 -- Brian Campbell <bacam@tardis.ed.ac.uk>  Sat,  9 Dec 2006 20:07:00 +0000



logmerge.ml
type ('a,'b) maybe = Found of 'a | NotFound of 'b

(* Return channels for all the files up to the one we've got the offset for. *)
let rec tryoffset variations filename inode offset =
  match variations with
    [] -> (prerr_endline ("No file for " ^ filename ^
                          " found which matches the last offset");
           [(NotFound offset, filename, inode)])
  | v::t ->
      try
        let file = open_in (filename ^ v) in
        let stat = Unix.fstat (Unix.descr_of_in_channel file) in
        let res = (Found file, filename, stat.Unix.st_ino) in
        if stat.Unix.st_ino = inode then (LargeFile.seek_in file offset; [res])
        else res::(tryoffset t filename inode offset)
      with Sys_error s ->
        (prerr_endline ("Unable to open file " ^ filename ^ v ^ ": " ^ s);
         tryoffset t filename inode offset)
;;

let openfiles offsetsfile =
  let rec readmore () =
    try
      let line = input_line offsetsfile in
      let this = Scanf.sscanf line "%S %d %Ld" (tryoffset [""; ".0"; ".1"]) in
      this::(readmore ())
    with End_of_file -> []
       | Scanf.Scan_failure s -> failwith ("offsets file corrupt: " ^ s)
  in readmore ()
;;

(* Pulls out times as an integer which respects order, more or less *)
let extracttime s =
  let themonth _ = 0 in (* FIX ME; need to cope with wrap arounds ? *)
  let convert month day hour minute second =
    second + 60*(minute + 60*(hour + 24*(day + 31*(themonth month)))) in
  Scanf.sscanf s "%s %d %d:%d:%d " convert  (* FIX ME - handle failure *)
;;

let rec insert ((lt,_,_) as l) lines =
  match lines with
    [] -> [l]
  | ((ht,_,_) as h)::t -> if lt < ht then l::lines
                                     else h::(insert l t)

let rec insertnewline src lines =
  match src with
    [] -> lines
      (* Actually, this should be the last element anyway. *)
  | (NotFound _,_,_)::t -> insertnewline t lines
  | (Found h,_,_)::t ->
      try
        let newline = input_line h in
        insert (extracttime newline, newline, src) lines
      with End_of_file -> insertnewline t lines

let nextline lines =
  match lines with
    [] -> []
  | (_,line,src)::t -> 
      (print_endline line; insertnewline src t)

let rec readfirstlines sources =
  match sources with
    [] -> []
  | h::t -> insertnewline (List.rev h) (readfirstlines t)

let merge sources =
  let lines = readfirstlines sources in
  let rec aux ls = match ls with [] -> () | _ -> aux (nextline ls) in
  aux lines

let savenewoffsets offsetsfile sources =
  let formatter = Format.formatter_of_out_channel offsetsfile in
  let writeoffset files =
    match files with
      [] -> prerr_endline "Warning: no files for a source!\n"
    | (NotFound offset,filename,inode)::_ ->
        Format.fprintf formatter "%S %d %Ld\n%!" filename inode offset
    | (Found file,filename,inode)::_ ->
        let offset = LargeFile.pos_in file in
        Format.fprintf formatter "%S %d %Ld\n%!" filename inode offset
  in List.iter writeoffset sources
;;

let offsetsfn = ref "/var/lib/logmerge/offsets";;
let update = ref true;;

let main () =
  Arg.parse [
      ("--no-update", Arg.Clear update, "Don't update offsets file");
      ("--offsets", Arg.Set_string offsetsfn, "File containing logfile offsets")
    ] (fun _ -> raise (Arg.Bad "Too many arguments"))
    "Usage: logmerge [--no-update] [--offsets <file>]";
  let offsetsfd = Unix.openfile !offsetsfn [Unix.O_RDWR] 0x600 in
  let offsetsfile = Unix.in_channel_of_descr offsetsfd in
  Unix.lockf offsetsfd Unix.F_TLOCK 0;
  let newoffsetsfn   = (!offsetsfn) ^ ".new" in
  let newoffsetsfile =
    if (!update) then Some (open_out newoffsetsfn) else None
  in
  let files = openfiles offsetsfile in
  merge files;
  match newoffsetsfile with
    None -> ()
  | Some newfile ->
    begin
      savenewoffsets newfile files;
      close_out newfile;
      Unix.rename newoffsetsfn (!offsetsfn);
    end;
  Unix.lockf offsetsfd Unix.F_ULOCK 0;
  close_in offsetsfile
in Unix.handle_unix_error main ();;
